gossip/src/crds.rs

```diff
@@ -789,39 +789,54 @@ mod tests {
     #[test]
     fn test_update_timestamp() {
         let mut crds = Crds::default();
-        let val = CrdsValue::new_unsigned(CrdsData::ContactInfo(ContactInfo::new_localhost(
+        let val1 = CrdsValue::new_unsigned(CrdsData::ContactInfo(ContactInfo::new_localhost(
             &Pubkey::default(),
             0,
         )));
+        let val1_hash = hash(&serialize(&val1).unwrap());
         assert_eq!(
-            crds.insert(val.clone(), 0, GossipRoute::LocalMessage),
+            crds.insert(val1.clone(), 0, GossipRoute::LocalMessage),
             Ok(())
         );
-        assert_eq!(crds.table[&val.label()].ordinal, 0);
+        assert_eq!(crds.table[&val1.label()].ordinal, 0);
+
+        // We want `val2` to overwrite `val1` based just on the hash value.  So, we look for a
+        // `ContactInfo` with a suitable hash.
+        let (val2, val2_hash, val2_gossip_port) = {
+            let mut port = 0;
+            loop {
+                let ci = ContactInfo::new_localhost_with_gossip_port(&Pubkey::default(), 0, port);
+                let val2 = CrdsValue::new_unsigned(CrdsData::ContactInfo(ci));
+                let val2_hash = hash(&serialize(&val2).unwrap());
+                if val2_hash > val1_hash {
+                    break (val2, val2_hash, port);
+                }
+                port = port.checked_add(1).expect("Can find a suitable hash value");
+            }
+        };
 
-        let val2 = CrdsValue::new_unsigned(CrdsData::ContactInfo(ContactInfo::default()));
-        let value_hash = hash(&serialize(&val2).unwrap());
-        assert_eq!(val2.label().pubkey(), val.label().pubkey());
+        assert_eq!(val2.label().pubkey(), val1.label().pubkey());
         assert_eq!(
             crds.insert(val2.clone(), 0, GossipRoute::LocalMessage),
             Ok(())
         );
```

Comment:
I think instead of relying on hash-value override for this insert to succeed, it would be better if `val2` is constructed with a later wallclock than the previous value. Similar to `ci.wallclock += 1;` in line 821 below.

The hash-value comparison in the override logic is only a last-ditch tie breaker which in practice should never occur; i.e. a node won't create two crds values with the same key within the same millisecond. But theoretically should that happen we need to tie break in a consistent and deterministic way across cluster, so we do the hash value comparison.

However, the wallclock override is the realistic scenario, and what we care about.

Reply:
Wouldn't it make the third check redundant?
I've removed hash based check and combined `val2` and `val3` assertions.
But would like to make sure that this is right.
As I am effectively removing coverage, then.

Reply:
I was suggesting something more like below.
Either way this will probably need to change again when we migrate from `LegacyContactInfo`, so maybe not worth spending a lot of time on it. 

```diff
diff --git a/gossip/src/crds.rs b/gossip/src/crds.rs
index 8e04012998..fe6bc294c0 100644
--- a/gossip/src/crds.rs
+++ b/gossip/src/crds.rs
@@ -801,7 +801,11 @@ mod tests {
         );
         assert_eq!(crds.table[&val.label()].ordinal, 0);
 
-        let val2 = CrdsValue::new_unsigned(CrdsData::LegacyContactInfo(ContactInfo::default()));
+        let val2 = {
+            let mut node = ContactInfo::default();
+            node.wallclock += 1;
+            CrdsValue::new_unsigned(CrdsData::LegacyContactInfo(node))
+        };
         let value_hash = hash(&serialize(&val2).unwrap());
         assert_eq!(val2.label().pubkey(), val.label().pubkey());
         assert_eq!(
@@ -820,7 +824,7 @@ mod tests {
         assert_eq!(crds.table[&val.label()].ordinal, 1);
 
         let mut ci = ContactInfo::default();
-        ci.wallclock += 1;
+        ci.wallclock += 2;
         let val3 = CrdsValue::new_unsigned(CrdsData::LegacyContactInfo(ci));
         assert_eq!(crds.insert(val3, 3, GossipRoute::LocalMessage), Ok(()));
         assert_eq!(*crds.purged.back().unwrap(), (value_hash, 3));
```

Reply:
Yes, if you make a change like this, then both updates are overwrites based on the `wallclock` value.
And there seems to be no value in doing it twice, no?

The second argument of `LegacyContactInfo::new_localhost()` is the `wallclock` value.
So using `LegacyContactInfo::new_localhost()` instead of `LegacyContactInfo::default()` allows me to skip explicit subsequent mutation.

> Either way this will probably need to change again when we migrate from LegacyContactInfo, so maybe not worth spending a lot of time on it.

True.  I'm not sure how much if this can be reused.
Hopefully most of the test?
In which case, there is probably some value in making it a bit easier to understand it :)